source/core/CEntryPoints.cpp

```diff
@@ -118,6 +118,46 @@ VIREO_EXPORT Int32 EggShell_PokeMemory(TypeManagerRef tm,
     }
 }
 //------------------------------------------------------------
+//! Allocates enough memory to fit a new object of TypeRef
+VIREO_EXPORT EggShellResult EggShell_AllocateData(TypeManagerRef tm, const TypeRef typeRef, void** dataRefLocation)
+{
+    TypeManagerScope scope(tm);
+    if (typeRef == nullptr || !typeRef->IsValid()) {
+        return kEggShellResult_InvalidTypeRef;
+    }
+
+    *dataRefLocation = nullptr;
+    Int32 topSize = typeRef->TopAQSize();
+    void* pData = THREAD_TADM()->Malloc(topSize);
+    NIError error = typeRef->InitData(pData, (TypeRef)nullptr);
+    if (error != kNIError_Success) {
+        return kEggShellResult_UnableToAllocateData;
+    }
+    *dataRefLocation = pData;
+    return kEggShellResult_Success;
+}
+//------------------------------------------------------------
+//! Deallocates data and frees up memory in dataRef described by typeRef
+VIREO_EXPORT EggShellResult EggShell_DeallocateData(TypeManagerRef tm, const TypeRef typeRef, void* dataRef)
+{
+    TypeManagerScope scope(tm);
+    if (typeRef == nullptr || !typeRef->IsValid()) {
+        return kEggShellResult_InvalidTypeRef;
+    }
+
+    if (dataRef == nullptr) {
+        return kEggShellResult_InvalidDataPointer;
+    }
+
+    NIError error = typeRef->ClearData(dataRef);
```

Comment:
Do we really need to clear the data before we free it?

Reply:
Yes.  Freeing only frees the top level, and knows nothing about the type.
ClearData deallocates any "hair" allocations in the type. For example, if a cluster contains a string, it would deallocate the string, and the free would deallocate the cluster top allocaiton.